exporter/otlphttpexporter/otlp.go

```diff
@@ -109,15 +122,78 @@ func (e *exporterImp) export(ctx context.Context, url string, request []byte) er
 
 	resp, err := e.client.Do(req)
 	if err != nil {
-		return fmt.Errorf("failed to push trace data via OTLP exporter: %w", err)
+		return fmt.Errorf("failed to make an HTTP request: %w", err)
+	}
+
+	defer func() {
+		// Discard any remaining response body when we are done reading.
+		io.CopyN(ioutil.Discard, resp.Body, maxHTTPResponseReadBytes)
+		resp.Body.Close()
+	}()
+
+	if resp.StatusCode >= 200 && resp.StatusCode <= 299 {
+		// Request is successful.
+		return nil
+	}
+
+	respStatus, err := readResponse(resp)
+
+	// Format the error message. Use the status if it is present in the response.
+	var formattedErr error
+	if err == nil && respStatus != nil {
+		formattedErr = fmt.Errorf(
+			"error exporting items, server responded with HTTP Status Code %d, Message=%s, Details=%v",
+			resp.StatusCode, respStatus.Message, respStatus.Details)
+	} else {
+		formattedErr = fmt.Errorf(
+			"error exporting items, server responded with HTTP Status Code %d",
+			resp.StatusCode)
 	}
 
-	_ = resp.Body.Close()
-	if resp.StatusCode < 200 || resp.StatusCode > 299 {
-		// TODO: Parse status and decide if can or not retry.
-		// TODO: Parse status and decide throttling.
-		return fmt.Errorf("failed the request with status code %d", resp.StatusCode)
+	// Check if the server is overwhelmed.
+	// See spec https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/protocol/otlp.md#throttling-1
+	if resp.StatusCode == http.StatusTooManyRequests || resp.StatusCode == http.StatusServiceUnavailable {
+		// Fallback to 0 if the Retry-After header is not present. This will trigger the
+		// default backoff policy by our caller (retry handler).
+		retryAfter := 0
+		if val := resp.Header.Get(headerRetryAfter); val != "" {
+			if seconds, err2 := strconv.Atoi(val); err2 == nil {
+				retryAfter = seconds
+			}
+		}
+		// Indicate to our caller to pause for the specified number of seconds.
+		return exporterhelper.NewThrottleRetry(formattedErr, time.Duration(retryAfter)*time.Second)
+	}
+
+	// Report the failure as permanent if the server thinks the request is malformed.
+	if resp.StatusCode == http.StatusBadRequest {
+		return consumererror.Permanent(formattedErr)
+	}
+
+	return formattedErr
+}
+
+func readResponse(resp *http.Response) (*status.Status, error) {
```

Comment:
It doesn't seem that you do anything with the returned error, you can simply return just the status.

Reply:
Good point.

Reply:
Fixed.